articles/113_ros_time.md

```diff
@@ -0,0 +1,115 @@
+Running faster than realtime can be valuable for high level testing as well allowing for repeated system tests.
+Slower than realtime simulation is necessary for complicated systems where accuracy is more important than speed.
+Often the simulation is the limiting factor for the system and as such the simulator can be a time source for faster or slower playback.
+And if the simulation is paused the system can also pause using the same mechanism.
+
+### Challenges in using abstracted time
+
+There are many algorithms for synchronization and they can typically achieve accuracies which are better than the latency of the network communications between devices on the network.
+However, these algorithms take advantage of assumptions about the constant and continuous nature of time.
+
+An important aspect of using an abstracted time is to be able to manipulate time.
+In some cases, changing the speed dynamically for example speeding up or slowing down all the way to pause is important for debugging.
+
+The ability to support pausing time requires that we not assume that the time values are always increasing.
+
+When communicating the changes in time propagation the latencies in the communication network becomes a challenge.
+Any change in the time abstraction must be communicated to the other nodes in the graph, but will be subject to normal network communication latency.
+This inaccuracy is proportional to the latency of communications and also proportional to the increase in the realtime factor.
+As such if very accurate timestamping is required using the time abstraction is required it can be achieved by slowing down the realtime factor such that the communication latency is comparatively small.
+
+The final challenge is that to be useful for log file playback the time abstraction must be able to jump backwards in time.
+This is a very different behavior from most clocks, and requires developers using the time abstraction to make sure their algorithm can deal with the discontinuity.
+
+## System Behavior
+
+To implement the time abstraction the following approach will be used.
+
+The time abstraction can be published by one source on the `/clock` topic.
+The topic will contain the latest abstracted time for the ROS system.
+If a publisher exists for the topic it will override the system time when using the ROS time abstraction.
+If `/clock` is being published calls to the ROS time abstraction will return the latest time received from the `/clock` topic with a default of zero if nothing has been received.
+
+If the time on the clock jumps backwards a callback handler will be invoked and be required to complete before any calls to the ROS time abstraction report the new time.
+Calls that come in before that must block.
+The developer has the opportunity to register callbacks with the handler to clear any state from their system if necessary before time will be in the past.
```

Comment:
I think there needs to be more precise language here. It's pretty confusing/unclear as to what calls are coming in and which ones must block. It might benefit from some examples, which might seem redundant, but I think would help to be clearer.